docs/proposals/ldap-group-sync.md

```diff
@@ -0,0 +1,189 @@
+An `LDAPCachedMemberLister` will be used inside of the sync job to do the first step in the sync: locate the LDAP entries that correspond all groups to be synced or all relevant members in the LDAP server, cache that result and decode it to determine the members of each group.
+
+```go
+// LDAPCachedMemberLister lists the LDAP entries that correspond to the members of the given Group
+type LDAPCachedMemberLister interface {
+  ListMembers(commonFilter string) (members [][]*ldap.Entry, err error)
+}
+```
+#### Given Members as First-Class LDAP Entries
+If members are first-class LDAP entries, the `LDAPCachedFirstClassMemberLister` can simply query all LDAP entries returned from the LDAP request for the attribute that is used to denote group membership. A list of unique groups is made, then cached LDAP response entries are once again filtered by this attribute to create lists of LDAP member entries for each group.
+
+#### Given Members as LDAP Annotations on LDAP Groups
+If groups are first-class LDAP entries, then `LDAPCachedReferencedMemberLister` can query for all groups to be synced first. The generation of the LDAP filter for a specific group from the description of the LDAP group in the LDAP group entry will need to be specified by the user requesting a sync. If LDAP group members are returned by reference in an LDAP entry attribute, another LDAP filter will need to be created with the given information in order to find the LDAP first-class entry for the user, if it exists. The generation of the LDAP filter for a specific member from the description of the LDAP group member in the LDAP group member entry will need to be specified by the user requesting a sync. Caching can be implemented for LDAP responses representing first-class members or the `LDAPMemberLister` can be used for each group once the LDAP unique filter is created for it.
+
+### Collecting LDAP Members: Searching for one Group
+An `LDAPMemberLister` will be used inside of the sync job to do the first step in the sync: locate the LDAP entries that correspond to the group being synced, given the filter for that group:
+
+```go
+// LDAPMemberLister lists the LDAP entries that correspond to the members of the given Group
+type LDAPMemberLister interface {
+  ListMembers(groupFilter string) (members *[]ldap.Entry, err error)
+}
+```
+#### Given Members as First-Class LDAP Entries
+If members are first-class LDAP entries, the `LDAPFirstClassMemberLister` can simply query the LDAP server for members of a group and return those entries.
+
+#### Given Members as LDAP Annotations on LDAP Groups
+If LDAP group members are returned by reference in an LDAP entry attribute, another LDAP filter will need to be created with the given information in order to find the LDAP first-class entry for the user, if it exists. The generation of the LDAP query from the description of the LDAP group member will need to be specified by the user requesting a sync.
```

Comment:
I accept the need to specify this.  I think it should be done in the config options.  I'm fine with saying that this command can only be run with an external config file.